core/src/main/java/org/ehcache/events/CacheEventDispatcherImpl.java

```diff
@@ -95,157 +130,115 @@ public void registerCacheEventListener(CacheEventListener<? super K, ? super V>
-        boolean interrupted = false;
-        try {
-          f.get();
-        } catch (ExecutionException e) {
-            LOGGER.error("Cache Event Listener failed for event type {} due to ", type, e);
-          // XXX delegate to resilience strategy (#52), and/or just log
-        } catch (InterruptedException e) {
-          interrupted = true;
-        } finally {
-          if (interrupted) {
-            Thread.currentThread().interrupt();
-          }
-        }
-      }
+    for (EventListenerWrapper listenerWrapper : syncListenersList) {
+      removeWrapperFromList(listenerWrapper, syncListenersList);
     }
+    orderedExecutor.shutdown();
   }
 
-  /**
-   * @return true if at least one cache event listener is registered
-   */
   @Override
-  public boolean hasListeners() {
-    return !registeredListeners.isEmpty();
+  public void setListenerSource(Cache<K, V> source) {
+    this.listenerSource = source;
   }
 
-  private final Set<EventListenerWrapper> registeredListeners = new CopyOnWriteArraySet<EventListenerWrapper>();
-  private final ExecutorService orderedDelivery;
-  private final ExecutorService unorderedDelivery;
-
-  private static final class EventListenerWrapper {
-    final CacheEventListener<?, ?> listener;
-    final CacheEventListenerConfiguration config;
-    
-    EventListenerWrapper(CacheEventListener<?, ?> listener, final EventFiring firing, final EventOrdering ordering, 
-        final EnumSet<EventType> forEvents) {
-      this.listener = listener;
-      this.config = new CacheEventListenerConfiguration() {
-        
-        @Override
-        public Class<CacheEventListenerProvider> getServiceType() {
-          return CacheEventListenerProvider.class;
-        }
-        
-        @Override
-        public EventOrdering orderingMode() {
-          return ordering;
-        }
-        
-        @Override
-        public EventFiring firingMode() {
-          return firing;
-        }
-        
-        @Override
-        public EnumSet<EventType> fireOn() {
-          return forEvents;
-        }
-      };
+  public void onEvent(CacheEvent<K, V> event) {
+    ExecutorService executor;
+    if (storeEventSource.isEventOrdering()) {
+      executor= orderedExecutor;
+    } else {
+      executor = unOrderedExectuor;
     }
-    
-    @SuppressWarnings("unchecked")
-    <K, V> CacheEventListener<K, V> getListener() {
-      return (CacheEventListener<K, V>) listener;
+    Future<?> future = null;
+    if (!syncListenersList.isEmpty()) {
+      future = executor.submit(new EventDispatchTask<K, V>(event, syncListenersList));
     }
-
-    @Override
-    public int hashCode() {
-      return listener.hashCode();
+    if (!aSyncListenersList.isEmpty()) {
+      executor.submit(new EventDispatchTask<K, V>(event, aSyncListenersList));
     }
-    
-    @Override
-    public boolean equals(Object other) {
-      if (!(other instanceof EventListenerWrapper)) {
-        return false;
+    if (future != null) {
```

Comment:
Can we avoid the null check here by pushing the future de-ref inside the if, and make associated scope change too.